### PR TITLE
Add registrable middleware

### LINKED ISSUE #320: Chihaya as a library
It should be possible to implement your own middleware and plug it into Chihaya without forking the repo. This mainly comes down to config parsing.

### PR DESCRIPTION
**Pull request by jzelinskie**
Fixes #320.

### CODE REVIEW

File: cmd/chihaya/main.go [hunk 1]
```diff
@@ -167,51 +172,61 @@ func RunCmdFunc(cmd *cobra.Command, args []string) error {
...
+
+	return nil
+}
+
+func main() {
+	var rootCmd = &cobra.Command{
+		Use:                "chihaya",
+		Short:              "BitTorrent Tracker",
+		Long:               "A customizable, multi-protocol BitTorrent Tracker",
+		PersistentPreRunE:  PreRunCmdFunc,
```
Comment (mrd0ll4r) [inline cmd/chihaya/main.go]: This is so much nicer!

File: middleware/clientapproval/clientapproval.go [hunk 2]
```diff
@@ -5,11 +5,34 @@ package clientapproval
...
+
+var _ middleware.Driver = driver{}
+
+type driver struct{}
+
+func (d driver) NewHook(optionBytes []byte) (middleware.Hook, error) {
+	var cfg Config
+	err := yaml.Unmarshal(optionBytes, &cfg)
+	if err != nil {
+		return nil, fmt.Errorf("invalid %s middleware: %s", Name, err)
```
Comment (mrd0ll4r) [inline middleware/clientapproval/clientapproval.go]: Maybe these should be more like `invalid options for middleware %s: %s` ?

File: middleware/logic.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/chihaya/chihaya/bittorrent"
+	"github.com/chihaya/chihaya/frontend"
+	"github.com/chihaya/chihaya/pkg/log"
+	"github.com/chihaya/chihaya/pkg/stop"
+	"github.com/chihaya/chihaya/storage"
+)
+
+// ResponseConfig holds the configuration used for the actual response.
+//
+// TODO(jzelinskie): Evaluate whether we would like to make this optional.
```
Comment (mrd0ll4r) [inline middleware/logic.go]: How can we make this optional and still have a working tracker?

File: middleware/jwt/jwt.go [hunk 4]
```diff
@@ -20,13 +21,35 @@ import (
 	"github.com/SermoDigital/jose/jws"
 	"github.com/SermoDigital/jose/jwt"
 	"github.com/mendsley/gojwk"
+	yaml "gopkg.in/yaml.v2"
```
Comment (mrd0ll4r) [inline middleware/jwt/jwt.go]: Something tells me you don't need the `yaml` rename here, can you check that please?

File: middleware/logic.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		if ctx, err = h.HandleScrape(ctx, req, resp); err != nil {
+			log.Error("post-scrape hooks failed", log.Err(err))
+			return
+		}
+	}
+}
+
+// Stop stops the Logic.
+//
+// This stops any hooks that implement stop.stop.
```
Comment (mrd0ll4r) [inline middleware/logic.go]: Nitpick: It actually implements `stop.Stopper`... sorry

File: middleware/jwt/jwt.go (hunk 4 above)
Comment (jzelinskie) [reply middleware/jwt/jwt.go]: Yup, I'll fix it. `goimports` did this for some reason.

File: middleware/logic.go (hunk 3 above)
Comment (jzelinskie) [reply middleware/logic.go]: That's the trade-off. We can make Chihaya extensible enough that you can program a new response generator at the cost of making config that won't compose a functional tracker.

Comment (mrd0ll4r) [reply middleware/logic.go]: Oooh, that's what you mean! Yeah, makes sense.

**Comment by jzelinskie** [comment]
@mrd0ll4r ptal

**Pull request merged**